### ISSUE TITLE
SessionBoundModel class

### ISSUE THREAD

**Issue by guillaumepotier** [tommikaikkonen/redux-orm#145]
Hi there,

I'm facing in my app a recurring problem: I quite often test my objects with `instanceof`, and I'm quite unable to do a `myobject instanceof SessionBoundModel` to ensure I properly have a `redux-orm` object in hand.

Currently, this is a one-liner `const` in Session file.

Do you think it would be possible to export it more globally? If so, I'd be glad to do a PR.

Thanks

**Comment by NathanBWaters** [comment]
@guillaumepotier I don't see anything wrong with that.  Stoked to see the PR!

**Comment by guillaumepotier** [comment]
Hi @NathanBWaters, having your official "go" here, I started digging deeper in the code and have some questions. This is not "as easy" as I initially thought (always that in dev^^) as `SessionBoundModel` dynamically extends each user defined models. I think I cannot just go away with declaring and export some static `SessionBoundModel` class right?

My 1rst question is: what is the real goal of that class anyway? Can't we just add `defineProperty` directly on the user Model and call `connect` on it too?

My 2nd question is: If we really want to have an interface for type checking (`instanceof SessionBoundModel` I do not see a way othen then defining a Proxy class. (Maybe we should implement it since ES6 `Proxy()` not supported by IE11 yet)

Best

**Comment by haveyaseen** [comment]
@guillaumepotier If we were to call `Model.connect` directly, we would permanently connect the `Model` to the session once we start a session of *any kind*, which makes it impossible to have simultaneous sessions. And we *do* have simultaneous sessions: [`memoize` starts a new session](https://github.com/tommikaikkonen/redux-orm/blob/v0.10.0/src/memoize.js#L155) while the user's session could still be running, which it probably is in 99.9% of the cases.

To come to your original point, could you please clarify the use-case for testing that a model is session bound? This is really library internal code and I don't see the need to expose it. Actually, you need to start a session in order to test that a Model is actually session bound, and by then you will already have the `session` object.

In other words: if you want to test that `session.YourModel` is an instance of `SessionBoundModel`, you could just as easily check that `session.YourModel.session` is present, which is quite superfluous because the library [already tests exactly this](https://github.com/tommikaikkonen/redux-orm/blob/v0.10.0/src/test/testSession.js#L35).

**Comment by guillaumepotier** [comment]
> In other words: if you want to test that session.YourModel is an instance of SessionBoundModel, you could just as easily check that session.YourModel.session is present, which is quite superfluous because the library already tests exactly this.

This could perfectly work for me, never thought of that ;)

I have many use cases needing to check that I'm correctly manipulating an orm resource object. I for example have a voter (checking acls) on some entities, that have this kind of api: `voter(resource, action)` and I would appreciate inside to directly test if `resource` is a proper orm entity to return `false` or throw an exception.

**Comment by haveyaseen** [comment]
Sounds reasonable, however you're probably just as well off testing `resource instanceof YourModel` in that case. Of course you can also check that `resource.session instanceof Session`. May or may not be necessary depending on how you start your session(s). It is kind of weird though as you can expect this to be the case when you're acting on a `resource` that stems from a `session` related query. :slightly_smiling_face: 

If this fixes your issue, please close it so we can reduce the number of issues.

**Comment by guillaumepotier** [comment]
Yup,

While re-reading this issue and comments, I'll definitely will test  `resource.session instanceof Session` or extend the same base model with another specific property for my resources.

Thanks for your time, great lib btw.

Best
